cmd/dep/status.go

```diff
@@ -643,3 +648,140 @@ func collectConstraints(ptree pkgtree.PackageTree, p *dep.Project, sm gps.Source
 	// TODO
 	return map[string][]gps.Constraint{}
 }
+
+type projectStatus struct {
+	Project               string
+	Version               string
+	Constraints           []string
+	Source                string
+	AltSource             string
+	PubVersions           map[string][]string
+	Revision              string
+	LatestAllowed         string
+	SourceType            string
+	Packages              []string
+	ProjectImporters      map[string]string
+	PackageImporters      map[string]string
+	UpstreamExists        bool
+	UpstreamVersionExists bool
+}
+
+func (ps projectStatus) String() string {
+	return fmt.Sprintf(`
+PROJECT:			%s
+VERSION:			%s
+CONSTRAINTS:			%s
+SOURCE:				%s
+ALT SOURCE:			%s
+PUB VERSION:			%s
+REVISION:			%s
+LATEST ALLOWED:			%s
+SOURCE TYPE:			%s
+PACKAGES:			%s
+PROJECT IMPORTERS:		%s
+PACKAGE IMPORTERS:		%s
+UPSTREAM EXISTS:		%t
+UPSTREAM VERSION EXISTS:	%t
+`,
+		ps.Project, ps.Version, ps.Constraints, ps.Source, ps.AltSource,
+		ps.PubVersions, ps.Revision, ps.LatestAllowed, ps.SourceType,
+		strings.Join(ps.Packages, ", "),
+		ps.ProjectImporters, ps.PackageImporters, ps.UpstreamExists,
+		ps.UpstreamVersionExists)
+}
+
+func runProjectStatus(ctx *dep.Ctx, args []string, p *dep.Project, sm gps.SourceManager) error {
+	// Collect all the project roots from the arguments.
+	var prs []gps.ProjectRoot
+
+	for _, arg := range args {
+		pr, err := sm.DeduceProjectRoot(arg)
+		if err != nil {
+			return err
+		}
+		prs = append(prs, pr)
+	}
+
+	for _, pr := range prs {
+		// Create projectStatus and add Project name.
+		projStatus := projectStatus{
+			Project: string(pr),
+		}
+
+		// Get the currently selected version from from lock.
+		for _, pl := range p.Lock.Projects() {
+			if pr == pl.Ident().ProjectRoot {
+				// VERSION
+				projStatus.Version = pl.Version().String()
+				// SOURCE
+				projStatus.Source = projStatus.Project
+				// ALT SOURCE
+				projStatus.AltSource = pl.Ident().Source
+
+				rev, _, _ := gps.VersionComponentStrings(pl.Version())
+				// REVISION
+				projStatus.Revision = rev
+
+				vcsType, err := sm.GetVcsType(pl.Ident())
+				if err != nil {
+					return err
+				}
+				// SOURCE TYPE
+				projStatus.SourceType = vcsType
+
+				existsUpstream, err := sm.ExistsUpstream(pl.Ident())
+				if err != nil {
+					return err
+				}
+				// UPSTREAM EXISTS
+				projStatus.UpstreamExists = existsUpstream
+
+				// Update local copy of the source and then fetch all the versions.
+				sm.SyncSourceFor(pl.Ident())
+				pvs, err := sm.ListVersions(pl.Ident())
```

Comment:
in the current implementation, `ListVersions()` is already guaranteed to always return the latest set of versions, so the `SyncSourceFor()` call is unnecessary.

that may change when the caching work goes in, but we don't need to code defensively against that eventuality - we'll cross that bridge when we come to it.